Add new metrics to help operators figure out whether there any stuck packets in a channel

adizere opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: #2175 

## Description

TODOs:

- [x] add mechanism workaround to handle `PosOverflow` errors in parse.
    - [ref in code](https://github.com/informalsystems/ibc-rs/compare/metrics_improvs?expand=1#diff-af85a6384e56bf271a402a5e5b9af4849f2b589ad9fe77f79adf68b1272daf73R24)
- [x] clarify and fix this comment https://github.com/informalsystems/ibc-rs/pull/2182#discussion_r878261892
- [X] clarify and fix this https://github.com/informalsystems/ibc-rs/pull/2182#discussion_r889001055
- [x] fix for reporting cf. #2175
______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules). 
- [x] Linked to GitHub issue.
- [x] Updated code comments and documentation (e.g., `docs/`).

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

adizere commented on relayer/src/worker/wallet.rs:
```diff
@@ -20,6 +20,9 @@ pub fn spawn_wallet_worker<Chain: ChainHandle>(chain: Chain) -> TaskHandle {
         })?;
 
+        // The input domain `balance.amount` may exceed u64::MAX.
+        // TODO: add mechanism workaround to handle `PosOverflow` errors in parse.
```
@romac when you get a chance, can you have a look at how we can fix this TODO please?

adizere commented on relayer/src/worker/wallet.rs:
```diff
@@ -20,6 +20,9 @@ pub fn spawn_wallet_worker<Chain: ChainHandle>(chain: Chain) -> TaskHandle {
         })?;
 
+        // The input domain `balance.amount` may exceed u64::MAX.
+        // TODO: add mechanism workaround to handle `PosOverflow` errors in parse.
```
@romac when you get a chance, can you have a look at how we can fix this TODO please?

romac replied:
https://github.com/informalsystems/ibc-rs/pull/2250/commits/d2ead45f90fbaac7941cf3789a8aa02a50a02771

romac commented:
@adizere Anything I can help with here?

adizere commented:
> @adizere Anything I can help with here?

Actually @ljoss17 is leading this effort, he has some interesting WIP in his local copy of this branch.

adizere commented:
I ran some tests and noticed we might have a possibly duplicate metric:

```
# HELP send_packet_count Number of SendPacket relayed.
# TYPE send_packet_count counter
send_packet_count{chain="ibc-0",channel="channel-0",counterparty="ibc-1",port="transfer"} 1101
send_packet_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 1610
```

which records data similarly to:

```
# HELP ibc_receive_packets Number of receive packets relayed per channel
# TYPE ibc_receive_packets counter
ibc_receive_packets{src_chain="ibc-0",src_channel="channel-0",src_port="transfer"} 1101
ibc_receive_packets{src_chain="ibc-1",src_channel="channel-0",src_port="transfer"} 1610
```

Looking in the code, it seems they are not the same, however:

1. `send_packet_count` is updated from the supervisor, using events directly received from the websocket.
    - this means that this metrics counts **all** send packet events that a chain emits (counting the activity of all relayers)
    - the metrics is also updated from packet clearing logic
2. `ibc_receive_packets` is updated from the packet worker
    - so this metrics counts _the_ send packet events that the **current Hermes instance has relayed** (not from all relayers active on that chain)

I think we introduced `send_packet_count` specifically for reporting on slow channels (#2175), but now that we have the `oldest_sequence` metric, if that counts accurately, then we can probably remove `send_packet_count`. Or should we keep it for some other reason?

ljoss17 commented:
> I ran some tests and noticed we might have a possibly duplicate metric:
> 
> ```
> # HELP send_packet_count Number of SendPacket relayed.
> # TYPE send_packet_count counter
> send_packet_count{chain="ibc-0",channel="channel-0",counterparty="ibc-1",port="transfer"} 1101
> send_packet_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 1610
> ```
> 
> which records data similarly to:
> 
> ```
> # HELP ibc_receive_packets Number of receive packets relayed per channel
> # TYPE ibc_receive_packets counter
> ibc_receive_packets{src_chain="ibc-0",src_channel="channel-0",src_port="transfer"} 1101
> ibc_receive_packets{src_chain="ibc-1",src_channel="channel-0",src_port="transfer"} 1610
> ```
> 
> Looking in the code, it seems they are not the same, however:
> 
> 1. `send_packet_count` is updated from the supervisor, using events directly received from the websocket.
>    
>    * this means that this metrics counts **all** send packet events that a chain emits (counting the activity of all relayers)
>    * the metrics is also updated from packet clearing logic
> 2. `ibc_receive_packets` is updated from the packet worker
>    
>    * so this metrics counts _the_ send packet events that the **current Hermes instance has relayed** (not from all relayers active on that chain)
> 
> I think we introduced `send_packet_count` specifically for reporting on slow channels (#2175), but now that we have the `oldest_sequence` metric, if that counts accurately, then we can probably remove `send_packet_count`. Or should we keep it for some other reason?

The `send_packet_count` could be useful if there are pending packets. The `oldest_sequence` will only give the oldest pending packet, while the `send_packet_count` combined with the `acknowledgement_count` will give how many packets are pending.
If this information doesn't seem very useful, then I don't see any reason to keep the `send_packet` and `acknowledgement_count`.
